File: CSharp/Controls/ReaderSettings/AdvancedReaderSettingsControl.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;
using Vintasoft.Barcode;
using Vintasoft.Barcode.BarcodeInfo;

namespace BarcodeDemo
{
    /// <summary>
    /// A control that allows to change the advanced settings of barcode reader.
    /// </summary>
    public partial class AdvancedReaderSettingsControl : ReaderSettingsEditorControl
    {

        #region Fields

        bool _enableSetSettings = true;

        #endregion



        #region Constructors

        /// <summary>
        /// Initializes a new instance of the <see cref="AdvancedReaderSettingsControl"/> class.
        /// </summary>
        public AdvancedReaderSettingsControl()
        {
            _enableSetSettings = false;

            InitializeComponent();

            barcodeReaderAustralianCustomInfoComboBox.Items.AddRange(EnumGetValues(typeof(AustralianPostCustomerInfoFormat)));

            barcodeReaderMSIChecksumComboBox.Items.Add(MSIChecksumType.Mod10);
            barcodeReaderMSIChecksumComboBox.Items.Add(MSIChecksumType.Mod11);
            barcodeReaderMSIChecksumComboBox.Items.Add(MSIChecksumType.Mod10Mod10);
            barcodeReaderMSIChecksumComboBox.Items.Add(MSIChecksumType.Mod11Mod10);
            barcodeReaderMSIChecksumComboBox.Items.Add(MSIChecksumType.None);

            toolTip1.SetToolTip(qrMaxAxialNonuniformityNumericUpDown, "QR maximum axial non-uniformity (x0.01)");

            _enableSetSettings = true;
        }

        #endregion



        #region Properties

        /// <summary>
        /// Gets or sets a value indicating whether barcode reader must interpret ECI characters.
        /// </summary>
        /// <value>
        /// <b>True</b> if barcode reader must interpret ECI characters; otherwise, <b>false</b>.
        /// </value>
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)]
        public bool InterpretEciCharacters
        {
            get
            {
                return interpretEciCharactersCheckBox.Checked;
            }
            set
            {
                interpretEciCharactersCheckBox.Checked = value;
            }
        }

        #endregion



        #region Methods

        private void settingCheckBox_CheckedChanged(object sender, EventArgs e)
        {
            SetSettings();
        }

        private void barcodeReaderAustralianCustomInfoComboBox_SelectedIndexChanged(
            object sender,
            EventArgs e)
        {
            SetSettings();
        }

        private void barcodeReaderMSIChecksumComboBox_SelectedIndexChanged(
            object sender,
            EventArgs e)
        {
            SetSettings();
        }

        private void DataMatrixLPatternMaxErasuresNumericUpDown_ValueChanged(
            object sender,
            EventArgs e)
        {
            SetSettings();
        }

        private void qrMaxANNumericUpDown_ValueChanged(object sender, EventArgs e)
        {
            SetSettings();
        }

        /// <summary>
        /// Updates the User Interface.
        /// </summary>
        public override void UpdateUI()
        {
            _enableSetSettings = false;

            optionalChecksumCheckBox.Checked = BarcodeReaderSettings.OptionalCheckSum;
            if (BarcodeReaderSettings.OptionalCheckSum)
                optionalChecksumCheckBox.ForeColor = Color.Red;
            else
                optionalChecksumCheckBox.ForeColor = Color.Black;
            searchDistortedDataMatrixBarcodesCheckBox.Checked = BarcodeReaderSettings.SearchDistortedDataMatrixBarcodes;
            searchDistortedQRBarcodesCheckBox.Checked = BarcodeReaderSettings.SearchDistortedQRBarcodes;
            telepenEnableNumericModeCheckBox.Checked = BarcodeReaderSettings.EnableTelepenNumericMode;
            collectQualityTestInformationCheckBox.Checked = BarcodeReaderSettings.CollectTestInformation;
            useCode128ExtendedDecodeTableCheckBox.Checked = BarcodeReaderSettings.Code128UseFnc4ShiftToExtendedAsciiTable;
            DataMatrixLPatternMaxErasuresNumericUpDown.Value = BarcodeReaderSettings.DataMatrixLPatternMaxErasures;
            barcodeCharacteristicSmallBarcodesCheckBox.Checked = (BarcodeReaderSettings.BarcodeCharacteristics & BarcodeCharacteristics.SmallBarcodes) != 0;
            barcodeCharacteristicNormalBarcodesCheckBox.Checked = (BarcodeReaderSettings.BarcodeCharacteristics & BarcodeCharacteristics.NormalSizeBarcodes) != 0;
            qrMaxAxialNonuniformityNumericUpDown.Value = (int)(BarcodeReaderSettings.QrMaxAxialNonuniformity * 100);
            recognitionTimeoutNumericUpDown.Value = BarcodeReaderSettings.RecognitionTimeout;
            code39WithoutStartStopCheckBox.Checked = BarcodeReaderSettings.SearchCode39WithoutStartStop;
            barcodeReaderAustralianCustomInfoComboBox.SelectedItem = BarcodeReaderSettings.AustralianPostCustomerInfoFormat;
            barcodeReaderMSIChecksumComboBox.SelectedItem = BarcodeReaderSettings.MSIChecksum;
            maxCellSizeNumericUpDown.Value = BarcodeReaderSettings.MatrixBarcodeMaxCellSize;
            searchQRModel1CheckBox.Checked = BarcodeReaderSettings.SearchQRModel1Barcodes;
            if (BarcodeReaderSettings.ImageScaleFactor <= 1 / 4f)
                downscale4RadioButton.Checked = true;
            else if (BarcodeReaderSettings.ImageScaleFactor <= 1 / 3f)
                downscale3RadioButton.Checked = true;
            else if (BarcodeReaderSettings.ImageScaleFactor <= 1 / 2f)
                downscale2RadioButton.Checked = true;
            else if (BarcodeReaderSettings.ImageScaleFactor <= 1.5f)
                noScaleRadioButton.Checked = true;
            else if (BarcodeReaderSettings.ImageScaleFactor <= 2f)
                upscale2RadioButton.Checked = true;
            else if (BarcodeReaderSettings.ImageScaleFactor <= 3f)
                upscale3RadioButton.Checked = true;
            invertColorsCheckBox.Checked = BarcodeReaderSettings.InvertImageColors;
            erodeCheckBox.Checked = BarcodeReaderSettings.Erode;
            removePeasCheckBox.Checked = BarcodeReaderSettings.RemovePeas;
            adaptiveBinarizationStepNumericUpDown.Value = BarcodeReaderSettings.AdaptiveBinarizationStep;

            _enableSetSettings = true;
        }

        /// <summary>
        /// Returns the enum values for specified type.
        /// </summary>
        private object[] EnumGetValues(Type type)
        {
            Array ar = Enum.GetValues(type);
            object[] result = new object[ar.Length];
            ar.CopyTo(result, 0);
            return result;
        }

        /// <summary>
        /// Sets the settings.
        /// </summary>
        private void SetSettings()
        {
            if (!_enableSetSettings)
                return;

            _enableSetSettings = false;

            BarcodeReaderSettings.OptionalCheckSum = optionalChecksumCheckBox.Checked;
            if (BarcodeReaderSettings.OptionalCheckSum)
                optionalChecksumCheckBox.ForeColor = Color.Red;
            else
                optionalChecksumCheckBox.ForeColor = Color.Black;

            BarcodeReaderSettings.SearchDistortedDataMatrixBarcodes = searchDistortedDataMatrixBarcodesCheckBox.Checked;
            BarcodeReaderSettings.DataMatrixLPatternMaxErasures = (int)DataMatrixLPatternMaxErasuresNumericUpDown.Value;
            BarcodeReaderSettings.SearchDistortedQRBarcodes = searchDistortedQRBarcodesCheckBox.Checked;
            BarcodeReaderSettings.EnableTelepenNumericMode = telepenEnableNumericModeCheckBox.Checked;
            BarcodeReaderSettings.CollectTestInformation = collectQualityTestInformationCheckBox.Checked;
            BarcodeReaderSettings.Code128UseFnc4ShiftToExtendedAsciiTable = useCode128ExtendedDecodeTableCheckBox.Checked;
            BarcodeReaderSettings.BarcodeCharacteristics = BarcodeCharacteristics.Undefinded;
            if (barcodeCharacteristicSmallBarcodesCheckBox.Checked)
                BarcodeReaderSettings.BarcodeCharacteristics |= BarcodeCharacteristics.SmallBarcodes;
            if (barcodeCharacteristicNormalBarcodesCheckBox.Checked)
                BarcodeReaderSettings.BarcodeCharacteristics |= BarcodeCharacteristics.NormalSizeBarcodes;

            BarcodeReaderSettings.AustralianPostCustomerInfoFormat = (AustralianPostCustomerInfoFormat)barcodeReaderAustralianCustomInfoComboBox.SelectedItem;
            BarcodeReaderSettings.MSIChecksum = (MSIChecksumType)barcodeReaderMSIChecksumComboBox.SelectedItem;
            try
            {
                BarcodeReaderSettings.RecognitionTimeout = (int)recognitionTimeoutNumericUpDown.Value;
            }
            catch (Exception e)
            {
                MessageBox.Show(e.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
            BarcodeReaderSettings.SearchCode39WithoutStartStop = code39WithoutStartStopCheckBox.Checked;
            BarcodeReaderSettings.MatrixBarcodeMaxCellSize = (int)maxCellSizeNumericUpDown.Value;
            BarcodeReaderSettings.QrMaxAxialNonuniformity = (float)qrMaxAxialNonuniformityNumericUpDown.Value / 100f;
            BarcodeReaderSettings.SearchQRModel1Barcodes = searchQRModel1CheckBox.Checked;
            if (downscale4RadioButton.Checked)
                BarcodeReaderSettings.ImageScaleFactor = 1 / 4f;
            else if (downscale3RadioButton.Checked)
                BarcodeReaderSettings.ImageScaleFactor = 1 / 3f;
            else if (downscale2RadioButton.Checked)
                BarcodeReaderSettings.ImageScaleFactor = 1 / 2f;
            else if (noScaleRadioButton.Checked)
                BarcodeReaderSettings.ImageScaleFactor = 1f;
            else if (upscale2RadioButton.Checked)
                BarcodeReaderSettings.ImageScaleFactor = 2f;
            else if (upscale3RadioButton.Checked)
                BarcodeReaderSettings.ImageScaleFactor = 3f;
            BarcodeReaderSettings.InvertImageColors = invertColorsCheckBox.Checked;
            BarcodeReaderSettings.Erode = erodeCheckBox.Checked;
            BarcodeReaderSettings.RemovePeas = removePeasCheckBox.Checked;
            BarcodeReaderSettings.AdaptiveBinarizationStep = (int)adaptiveBinarizationStepNumericUpDown.Value;

            _enableSetSettings = true;
        }

        /// <summary>
        /// Sets BarcodeReaderSettings.RecognitionTimeout value.
        /// </summary>
        private void recognitionTimeoutNumericUpDown_ValueChanged(object sender, EventArgs e)
        {
            try
            {
                BarcodeReaderSettings.RecognitionTimeout = (int)recognitionTimeoutNumericUpDown.Value;
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
        }

        /// <summary>
        /// Sets BarcodeReaderSettings.MatrixBarcodeMaxCellSize value.
        /// </summary>
        private void maxCellSizeNumericUpDown_ValueChanged(object sender, EventArgs e)
        {
            BarcodeReaderSettings.MatrixBarcodeMaxCellSize = (int)maxCellSizeNumericUpDown.Value;
        }

        /// <summary>
        /// Raises the <see cref="E:ImageProcessingSettingsChanged" /> event.
        /// </summary>
        private void OnImageProcessingSettingsChanged(EventArgs e)
        {
            if (ImageProcessingSettingsChanged != null)
                ImageProcessingSettingsChanged(this, e);
        }

        /// <summary>
        /// Set the reader settings.
        /// </summary>
        private void imageProcessing_SettingsChanged(object sender, EventArgs e)
        {
            SetSettings();
            OnImageProcessingSettingsChanged(EventArgs.Empty);
        }

        #endregion



        #region Events

        /// <summary>
        /// The image processing settings are changed.
        /// </summary>
        public event EventHandler ImageProcessingSettingsChanged;

        #endregion

    }
}

File: VB.NET/Controls/ReaderSettings/AdvancedReaderSettingsControl.vb
Imports System.ComponentModel
Imports System.Windows.Forms

Imports Vintasoft.Barcode
Imports Vintasoft.Barcode.BarcodeInfo
Imports System.Drawing

''' <summary>
''' Represents an editor control of advanced reader settings.
''' </summary>
Public Partial Class AdvancedReaderSettingsControl
	Inherits ReaderSettingsEditorControl

	#Region "Fields"

	Private _enableSetSettings As Boolean = True

	#End Region



	#Region "Constructors"

	Public Sub New()
		InitializeComponent()

		barcodeReaderAustralianCustomInfoComboBox.Items.AddRange(EnumGetValues(GetType(AustralianPostCustomerInfoFormat)))

		barcodeReaderMSIChecksumComboBox.Items.Add(MSIChecksumType.Mod10)
		barcodeReaderMSIChecksumComboBox.Items.Add(MSIChecksumType.Mod11)
		barcodeReaderMSIChecksumComboBox.Items.Add(MSIChecksumType.Mod10Mod10)
		barcodeReaderMSIChecksumComboBox.Items.Add(MSIChecksumType.Mod11Mod10)
		barcodeReaderMSIChecksumComboBox.Items.Add(MSIChecksumType.None)
	End Sub


	#End Region



	#Region "Properties"

	Private _barcodeTypesReaderSettings As BarcodeTypesReaderSettingsControl = Nothing
	Public Property BarcodeTypesReaderSettings() As BarcodeTypesReaderSettingsControl
        Get
            Return _barcodeTypesReaderSettings
        End Get
        Set(ByVal value As BarcodeTypesReaderSettingsControl)
            _barcodeTypesReaderSettings = Value
            UpdateUI()
        End Set
    End Property

	Public Property InterpretEciCharacters() As Boolean
		Get
			Return interpretEciCharactersCheckBox.Checked
		End Get
		Set
			interpretEciCharactersCheckBox.Checked = value
		End Set
	End Property

	#End Region



	#Region "Methods"

    Private Sub settingCheckBox_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles useCode128ExtendedDecodeTableCheckBox.CheckedChanged, telepenEnableNumericModeCheckBox.CheckedChanged, searchDistortedQRBarcodesCheckBox.CheckedChanged, searchDistortedDataMatrixBarcodesCheckBox.CheckedChanged, optionalChecksumCheckBox.CheckedChanged, collectQualityTestInformationCheckBox.CheckedChanged, barcodeCharacteristicSmallBarcodesCheckBox.CheckedChanged
        SetSettings()
    End Sub

    Private Sub barcodeReaderAustralianCustomInfoComboBox_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles barcodeReaderAustralianCustomInfoComboBox.SelectedIndexChanged
        SetSettings()
    End Sub

    Private Sub barcodeReaderMSIChecksumComboBox_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles barcodeReaderMSIChecksumComboBox.SelectedIndexChanged
        SetSettings()
    End Sub

    Private Sub DataMatrixLPatternMaxErasuresNumericUpDown_ValueChanged(ByVal sender As Object, ByVal e As EventArgs) Handles DataMatrixLPatternMaxErasuresNumericUpDown.ValueChanged
        SetSettings()
    End Sub

	Public Overrides Sub UpdateUI()
		_enableSetSettings = False

		optionalChecksumCheckBox.Checked = BarcodeReaderSettings.OptionalCheckSum
		If BarcodeReaderSettings.OptionalCheckSum Then
			optionalChecksumCheckBox.ForeColor = Color.Red
		Else
			optionalChecksumCheckBox.ForeColor = Color.Black
		End If
		searchDistortedDataMatrixBarcodesCheckBox.Checked = BarcodeReaderSettings.SearchDistortedDataMatrixBarcodes
		searchDistortedQRBarcodesCheckBox.Checked = BarcodeReaderSettings.SearchDistortedQRBarcodes
		telepenEnableNumericModeCheckBox.Checked = BarcodeReaderSettings.EnableTelepenNumericMode
		collectQualityTestInformationCheckBox.Checked = BarcodeReaderSettings.CollectTestInformation
		useCode128ExtendedDecodeTableCheckBox.Checked = BarcodeReaderSettings.Code128UseFnc4ShiftToExtendedAsciiTable
		DataMatrixLPatternMaxErasuresNumericUpDown.Value = BarcodeReaderSettings.DataMatrixLPatternMaxErasures
		barcodeCharacteristicSmallBarcodesCheckBox.Checked = (BarcodeReaderSettings.BarcodeCharacteristics And BarcodeCharacteristics.SmallBarcodes) <> 0

		barcodeReaderAustralianCustomInfoComboBox.SelectedItem = BarcodeReaderSettings.AustralianPostCustomerInfoFormat
		barcodeReaderMSIChecksumComboBox.SelectedItem = BarcodeReaderSettings.MSIChecksum

		If _barcodeTypesReaderSettings IsNot Nothing Then
			_barcodeTypesReaderSettings.HighlightOptionalChecksumBarcodes = optionalChecksumCheckBox.Checked
		End If

		_enableSetSettings = True
	End Sub

	Private Function EnumGetValues(type As Type) As Object()
		Dim ar As Array = [Enum].GetValues(type)
		Dim result As Object() = New Object(ar.Length - 1) {}
		ar.CopyTo(result, 0)
		Return result
	End Function

	Private Sub SetSettings()
		If Not _enableSetSettings Then
			Return
		End If

		_enableSetSettings = False

		BarcodeReaderSettings.OptionalCheckSum = optionalChecksumCheckBox.Checked
		If BarcodeReaderSettings.OptionalCheckSum Then
			optionalChecksumCheckBox.ForeColor = Color.Red
		Else
			optionalChecksumCheckBox.ForeColor = Color.Black
		End If

		BarcodeReaderSettings.SearchDistortedDataMatrixBarcodes = searchDistortedDataMatrixBarcodesCheckBox.Checked
		BarcodeReaderSettings.DataMatrixLPatternMaxErasures = CInt(Math.Truncate(DataMatrixLPatternMaxErasuresNumericUpDown.Value))
		BarcodeReaderSettings.SearchDistortedQRBarcodes = searchDistortedQRBarcodesCheckBox.Checked
		BarcodeReaderSettings.EnableTelepenNumericMode = telepenEnableNumericModeCheckBox.Checked
		BarcodeReaderSettings.CollectTestInformation = collectQualityTestInformationCheckBox.Checked
		BarcodeReaderSettings.Code128UseFnc4ShiftToExtendedAsciiTable = useCode128ExtendedDecodeTableCheckBox.Checked
		If barcodeCharacteristicSmallBarcodesCheckBox.Checked Then
			BarcodeReaderSettings.BarcodeCharacteristics = BarcodeCharacteristics.SmallBarcodes
		Else
			BarcodeReaderSettings.BarcodeCharacteristics = BarcodeCharacteristics.Undefinded
		End If

		BarcodeReaderSettings.AustralianPostCustomerInfoFormat = DirectCast(barcodeReaderAustralianCustomInfoComboBox.SelectedItem, AustralianPostCustomerInfoFormat)
		BarcodeReaderSettings.MSIChecksum = DirectCast(barcodeReaderMSIChecksumComboBox.SelectedItem, MSIChecksumType)

		If _barcodeTypesReaderSettings IsNot Nothing Then
			_barcodeTypesReaderSettings.HighlightOptionalChecksumBarcodes = optionalChecksumCheckBox.Checked
		End If

		_enableSetSettings = True
	End Sub

	#End Region

End Class
